src/libraries/Microsoft.Extensions.Configuration/tests/ConfigTests.cs

```diff
@@ -0,0 +1,1200 @@
+
+            var config = new Config();
+
+            ((IConfigurationBuilder)config).Add(new TestConfigurationSource(providerMock.Object));
+
+            Assert.NotEmpty(changeToken.Callbacks);
+
+            config.Dispose();
+
+            Assert.Empty(changeToken.Callbacks);
+        }
+
+        [Fact]
+        public void DisposesChangeTokenRegistrationsOnRemoval()
+        {
+            var changeToken = new TestChangeToken();
+            var providerMock = new Mock<IConfigurationProvider>();
+            providerMock.Setup(p => p.GetReloadToken()).Returns(changeToken);
+
+            var source = new TestConfigurationSource(providerMock.Object);
+
+            var config = new Config();
+            IConfigurationBuilder builder = config;
+
+            builder.Add(source);
+
+            Assert.NotEmpty(changeToken.Callbacks);
+
+            builder.Sources.Remove(source);
+
+            Assert.Empty(changeToken.Callbacks);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void ChainedConfigurationIsDisposedOnDispose(bool shouldDispose)
+        {
+            var provider = new DisposableTestConfigurationProvider("foo", "foo-value");
+            var chainedConfig = new ConfigurationRoot(new IConfigurationProvider[] {
+                provider
+            });
+
+            var config = new Config();
+
+            config.AddConfiguration(chainedConfig, shouldDisposeConfiguration: shouldDispose);
+
+            Assert.False(provider.IsDisposed);
+
+            config.Dispose();
+
+            Assert.Equal(shouldDispose, provider.IsDisposed);
+        }
+
+        [Fact]
+        public void LoadAndCombineKeyValuePairsFromDifferentConfigurationProviders()
+        {
+            // Arrange
+            var dic1 = new Dictionary<string, string>()
+            {
+                {"Mem1:KeyInMem1", "ValueInMem1"}
+            };
+            var dic2 = new Dictionary<string, string>()
+            {
+                {"Mem2:KeyInMem2", "ValueInMem2"}
+            };
+            var dic3 = new Dictionary<string, string>()
+            {
+                {"Mem3:KeyInMem3", "ValueInMem3"}
+            };
+            var memConfigSrc1 = new MemoryConfigurationSource { InitialData = dic1 };
+            var memConfigSrc2 = new MemoryConfigurationSource { InitialData = dic2 };
+            var memConfigSrc3 = new MemoryConfigurationSource { InitialData = dic3 };
+
+            var config = new Config();
+            IConfigurationBuilder configurationBuilder = config;
+
+            // Act
+            configurationBuilder.Add(memConfigSrc1);
+            configurationBuilder.Add(memConfigSrc2);
+            configurationBuilder.Add(memConfigSrc3);
+
+            var memVal1 = config["mem1:keyinmem1"];
+            var memVal2 = config["Mem2:KeyInMem2"];
+            var memVal3 = config["MEM3:KEYINMEM3"];
+
+            // Assert
```

Comment:
Nit: This comment doesn't seem that useful.

Reply:
I generally agree about the uselessness of `// Arrange // Act // Assert` comments, but this was copied from ConfigurationTest. If we change it, I'd rather do it in one pass.

Reply:
Ok, I don't mind if we do that cleanup in a separate PR.